### PR TITLE
Bug fix: input isn't valid multihash

### PR DESCRIPTION
**Pull request by kishansagathiya**
Some command were failing with `input isn't valid multihash`, due to
empty peer.ID. This commit omit peer id from IPFSID when empty, thus
avoiding the error state

#715

### CODE REVIEW

**Comment by hsanjuan** [comment]
It seems `Peers()` wraps errors with empty ID objects with the `Error` field set, but these cannot be re-parsed after convering to json because IPFSID.ID is empty? Is this the case? Probably the whole IPFSID object should be omitted in these cases because it does not make sense to marshal any part of it (might need to make it pointer and leave it nil).

There also the case when the IPFS daemon errors and IPFSID carries the error only. That will be fixed by the `omitempty` here. 

The `PinInfo` object also carries a `peer.ID` and an Error. We should make sure the `Peer` field is always set, even when it is used to wrap an error, or that `omitempty` is also used for that field.

**Comment by kishansagathiya** [comment]
> It seems Peers() wraps errors with empty ID objects with the Error field set, but these cannot be re-parsed after convering to json because IPFSID.ID is empty? Is this the case?

Yes

**Comment by kishansagathiya** [comment]
There isn't any problem because of peerID not being empty in `PinInfo`(may be because it is used at pointer at most places), at least in this scenario. I can `omitempty` it to be safe. But if we keep it we would know about those failure cases(when this should be omitted) instead of silently not showing peer ID.

File: ipfsconn/ipfshttp/ipfshttp.go [hunk 1]
```diff
@@ -512,6 +512,9 @@ func (ipfs *Connector) ConnectSwarms(ctx context.Context) error {
 
 	for _, id := range ids {
 		ipfsID := id.IPFS
+		if ipfsID == nil {
```
Comment (hsanjuan) [inline ipfsconn/ipfshttp/ipfshttp.go]: should `continue` if `id.Error != ""`, `id.IPFS == nil` (hopefully this should never happen), or `id.IPFS.Error != ""`

File: api/types.go [hunk 2]
```diff
@@ -344,14 +344,14 @@ type ID struct {
 	Commit                string      `json:"commit" codec:"c,omitempty"`
 	RPCProtocolVersion    protocol.ID `json:"rpc_protocol_version" codec:"rv,omitempty"`
 	Error                 string      `json:"error" codec:"e,omitempty"`
-	IPFS                  IPFSID      `json:"ipfs" codec:"ip,omitempty"`
+	IPFS                  *IPFSID     `json:"ipfs" codec:"ip,omitempty"`
```
Comment (hsanjuan) [inline api/types.go]: add `omitempty` to the json tag.

**Pull request merged**